### PR TITLE
[useAnimatedKeyboard][iOS] Keyboard interpolation

### PR DESCRIPTION
**Pull request by piaskowyk**
## Summary

The `useAnimatedKeyboard()` hook doesn't sync with the keyboard animation because keyboard animations use CAAnimations that are computed on the window server. This means they have their own timer and presentation layer. The info we get about the actual keyboard height is just an approximation of the real size shown by the window server.

Our hook, on the other hand, uses a DisplayLink timer, which is different from the CAAnimation timer. With this PR, we are adding a function that attempts to better estimate the keyboard position using easing and Bezier curves. While this solution might not be perfect down to the pixel because window server animations aren't entirely straightforward, it's still a big improvement compared to before.

The Bezier curves are based on the measured keyboard height from screen recordings, and then I adjusted the curves to fit them better.

Basic estimation function:
image

Where:
- x - time progress of animation
- a1, a2, b1, b2, d1, d3 - curve parameters

Example of curve:

image

Keyboard appearing: https://www.desmos.com/calculator/aw4ocksord
Keyboard hiding: https://www.desmos.com/calculator/6zay2c6jou

Related info: https://forums.developer.apple.com/forums/thread/712762

## The curve parameters

"The curve parameters were calculated using an optimization script that aimed to minimize approximation errors. The script consists of three stages:
- Analyzing the video to compute the keyboard frame during animations
- Calculating the average keyboard height during the animation
- Initial stage of curve approximation with a step of 0.1
- Additional stage of curve approximation with a step of 0.01"

Results:
best_factors_opening: 1, 4.88, 2.59, 9.89, 0.29, 2.45, error: 0.0149869186740412
best_factors_closing: 1, 5.65, 2.74, 8.38, 0.93, 3.29, error: 0.02679889950216266

<details>
<summary>code</summary>

```python
import cv2
import numpy as np

cap = cv2.VideoCapture('video.mp4')
if not cap.isOpened():
  print("Error opening video file")
  exit()

opening_measurements = []
closing_measurements = []
current_measurements = []

last_keyboard_height = 0
state = ""

while cap.isOpened():
  ret, frame = cap.read()
  if not ret:
    break
  frame_height = int(cap.get(cv2.CAP_PROP_FRAME_HEIGHT))
  frame_width = int(cap.get(cv2.CAP_PROP_FRAME_WIDTH))
  gray_frame = cv2.cvtColor(frame, cv2.COLOR_BGR2GRAY)
  _, binary_frame = cv2.threshold(gray_frame, 240, 255, cv2.THRESH_BINARY)

  keyboard_height = 0
  for i in range(frame_height - 1, 0, -1):
    if binary_frame[i][frame_width  - 1] == 255:
      keyboard_height = frame_height - i - 1;
      print(f"Keyboard height = {keyboard_height}")
      break
  
  if keyboard_height > last_keyboard_height or (keyboard_height > 1 and keyboard_height < 757):
    if (len(current_measurements) == 0):
      current_measurements.append(last_keyboard_height)
    current_measurements.append(keyboard_height)
    state = "opening"
  elif keyboard_height < last_keyboard_height or (keyboard_height > 1 and keyboard_height < 757):
    if (len(current_measurements) == 0):
      current_measurements.append(last_keyboard_height)
    current_measurements.append(keyboard_height)
    state = "closing"
  elif len(current_measurements) > 0:
    if state == "opening":
      opening_measurements.append(current_measurements)
    elif state == "closing":
      closing_measurements.append(current_measurements)
    current_measurements = []
  last_keyboard_height = keyboard_height

  cv2.imshow('Frame', binary_frame)
  if cv2.waitKey(1) & 0xFF == ord('q'):
    break

cap.release()
cv2.destroyAllWindows()

opening_by_frame = []
average_opening = []
deviation_opening = []
for i in range(len(opening_measurements[0])):
  opening_by_frame.append([])
for i in range(len(opening_measurements[0])):
  for measure in opening_measurements:
    opening_by_frame[i].append(measure[i])
for (i, measure) in enumerate(opening_by_frame):
  average_opening.append(np.mean(measure))
  deviation_opening.append(np.std(measure)))

closing_by_frame = []
average_closing = []
deviation_closing = []
for i in range(len(closing_measurements[0])):
  closing_by_frame.append([])
for i in range(len(closing_measurements[0])):
  for measure in closing_measurements:
    closing_by_frame[i].append(measure[i])
for (i, measure) in enumerate(closing_by_frame):
  average_closing.append(np.mean(measure))
  deviation_closing.append(np.std(measure)))

for (i, measure) in enumerate(opening_measurements):
  print(f"Opening {i}: {len(measure)}, {measure}")

for (i, measure) in enumerate(closing_measurements):
  print(f"Closing {i}: {len(measure)}, {measure}")

print(f"Average opening: {average_opening}")
print(f"Deviation opening: {deviation_opening}")
print(f"Average closing: {average_closing}")
print(f"Deviation closing: {deviation_closing}")


# average_opening = [0.0, 68.9090909090909, 146.0909090909091, 233.72727272727272, 313.6363636363636, 387.90909090909093, 454.0, 510.72727272727275, 558.7272727272727, 598.6363636363636, 632.0, 658.0909090909091, 678.9090909090909, 695.9090909090909, 709.3636363636364, 719.7272727272727, 728.2727272727273, 734.9090909090909, 739.9090909090909, 744.0, 747.0909090909091, 749.2727272727273, 751.1818181818181, 752.7272727272727, 753.7272727272727, 754.1818181818181, 755.0, 756.0, 756.0, 757.0]

# average_closing = [757.0, 674.8181818181819, 595.0909090909091, 508.1818181818182, 428.54545454545456, 355.90909090909093, 291.6363636363636, 236.72727272727272, 190.36363636363637, 152.36363636363637, 119.36363636363636, 95.0909090909091, 75.36363636363636, 58.81818181818182, 46.18181818181818, 36.45454545454545, 27.90909090909091, 21.636363636363637, 16.636363636363637, 13.0, 10.0, 7.454545454545454, 6.0, 4.090909090909091, 3.4545454545454546, 2.272727272727273, 2.1818181818181817, 2.0, 2.0, 1.0, 0.0]

for i in range(len(average_opening)):
  average_opening[i] = average_opening[i] / 757
  average_closing[i] = average_closing[i] / 757

def formula(x, a1, a2, b1, b2, c1, c2):
  return 1 - a1 * ((1 - x) ** a2) - b1 * x * ((1 - x) ** b2) - c1 * (x ** 2) * ((1 - x) ** c2)

iteration_counter = 0
best_error = 1000000
best_opening = ""
best_opening_factors = []
a1 = 1
for a2 in range(40, 55):
  a2 = a2 / 10
  for b1 in range(10, 35):
    b1 = b1 / 10
    for b2 in range(60, 100):
      b2 = b2 / 10
      for c1 in range(30):
        c1 = c1 / 10
        for c2 in range(10, 50):
          c2 = c2 / 10
          values = []
          for x in range(30):
            x = x / 30
            value = formula(x, a1, a2, b1, b2, c1, c2)
            values.append(value)
          error = 0
          for i in range(30):
            error += abs(values[i] - average_opening[i])
          if error < best_error:
            best_error = error
            best_opening_factors = [a1, a2, b1, b2, c1, c2]
            best_opening = f"best_factors_opening: {a1}, {a2}, {b1}, {b2}, {c1}, {c2}, error: {error}"
            print(best_opening)
          iteration_counter += 1
          if iteration_counter % 1000000 == 0:
            iteration_counter = 0
            print(f"current_factors: {a1}, {a2}, {b1}, {b2}, {c1}, {c2}, error: {error}, best_error: {best_error}")

iteration_counter = 0
best_error = 1000000
best_closing = ""
best_closing_factors = []
a1 = 1
for a2 in range(40, 75):
  a2 = a2 / 10
  for b1 in range(10, 45):
    b1 = b1 / 10
    for b2 in range(50, 85):
      b2 = b2 / 10
      for c1 in range(20):
        c1 = c1 / 10
        for c2 in range(20, 40):
          c2 = c2 / 10
          values = []
          for x in range(30):
            x = x / 30
            value = formula(x, a1, a2, b1, b2, c1, c2)
            values.append(value)
          error = 0
          for i in range(30):
            error += abs(values[i] - (1 - average_closing[i]))
          if error < best_error:
            best_error = error
            best_opening_factors = [a1, a2, b1, b2, c1, c2]
            best_closing = f"best_factors_closing: {a1}, {a2}, {b1}, {b2}, {c1}, {c2}, error: {error}"
            print(best_closing)
          iteration_counter += 1
          if iteration_counter % 1000000 == 0:
            iteration_counter = 0
            print(f"current_factors: {a1}, {a2}, {b1}, {b2}, {c1}, {c2}, error: {error}, best_error: {best_error}")


print("--------------------------------------")
print(best_opening)
print(best_closing)
print("--------------------------------------")

#####################################################################################

a1_copy, a2_copy, b1_copy, b2_copy, c1_copy, c2_copy = best_opening_factors

iteration_counter = 0
best_error = 1000000
best_opening = ""
a1 = 1
a2_best, b1_best, b2_best, c1_best, c2_best = a2_copy, b1_copy, b2_copy, c1_copy, c2_copy
for a2 in range(int((a2_best - 0.1) * 100), int((a2_best + 0.1) * 100)):
  a2 = a2 / 100
  for b1 in range(int((b1_best - 0.1) * 100), int((b1_best + 0.1) * 100)):
    b1 = b1 / 100
    for b2 in range(int((b2_best - 0.1) * 100), int((b2_best + 0.1) * 100)):
      b2 = b2 / 100
      for c1 in range(int((c1_best - 0.1) * 100), int((c1_best + 0.1) * 100)):
        c1 = c1 / 100
        for c2 in range(int((c2_best - 0.1) * 100), int((c2_best + 0.1) * 100)):
          c2 = c2 / 100
          values = []
          for x in range(30):
            x = x / 30
            value = formula(x, a1, a2, b1, b2, c1, c2)
            values.append(value)
          error = 0
          for i in range(30):
            error += abs(values[i] - average_opening[i])
          if error < best_error:
            best_error = error
            best_opening = f"best_factors_opening: {a1}, {a2}, {b1}, {b2}, {c1}, {c2}, error: {error}"
            print(best_opening)
          iteration_counter += 1
          if iteration_counter % 1000000 == 0:
            iteration_counter = 0
            print(f"current_factors: {a1}, {a2}, {b1}, {b2}, {c1}, {c2}, error: {error}, best_error: {best_error}")

iteration_counter = 0
best_error = 1000000
best_closing = ""
a1 = 1
a2_best, b1_best, b2_best, c1_best, c2_best = best_closing_factors
for a2 in range(int((a2_best - 0.1) * 100), int((a2_best + 0.1) * 100)):
  a2 = a2 / 100
  for b1 in range(int((b1_best - 0.1) * 100), int((b1_best + 0.1) * 100)):
    b1 = b1 / 100
    for b2 in range(int((b2_best - 0.1) * 100), int((b2_best + 0.1) * 100)):
      b2 = b2 / 100
      for c1 in range(int((c1_best - 0.1) * 100), int((c1_best + 0.1) * 100)):
        c1 = c1 / 100
        for c2 in range(int((c2_best - 0.1) * 100), int((c2_best + 0.1) * 100)):
          c2 = c2 / 100
          values = []
          for x in range(30):
            x = x / 30
            value = formula(x, a1, a2, b1, b2, c1, c2)
            values.append(value)
          error = 0
          for i in range(30):
            error += abs(values[i] - (1 - average_closing[i]))
          if error < best_error:
            best_error = error
            best_closing = f"best_factors_closing: {a1}, {a2}, {b1}, {b2}, {c1}, {c2}, error: {error}"
            print(best_closing)
          iteration_counter += 1
          if iteration_counter % 1000000 == 0:
            iteration_counter = 0
            print(f"current_factors: {a1}, {a2}, {b1}, {b2}, {c1}, {c2}, error: {error}, best_error: {best_error}")

print("--------------------------------------")
print(best_opening)
print(best_closing)
print("--------------------------------------")
```

</details>

## Comparison

| before | after |
| --- | --- |
| <video src="https://github.com/software-mansion/react-native-reanimated/assets/36106620/05f7936e-dc99-4ab6-a59f-39e9c3c2ed82" /> | <video src="https://github.com/software-mansion/react-native-reanimated/assets/36106620/004a2c5e-6925-4d26-8cfb-3f098702c9d0" /> |

## Test plan

<details>
<summary>code</summary>

```js
import React from 'react';
import { Text, StyleSheet, View, ScrollView, TextInput } from 'react-native';
import Animated, { useAnimatedKeyboard, useAnimatedStyle } from 'react-native-reanimated';

export default function EmptyExample() {
  const keyboard = useAnimatedKeyboard();
  const translateStyle = useAnimatedStyle(() => {
    return {
      transform: [{ translateY: -keyboard.height.value }],
    };
  });
  return (
    <View style={styles.container}>
      <ScrollView 
        style={styles.scroll}
        keyboardDismissMode='interactive'
      >
        {new Array(100).fill(null).map((_, i) => <Text key={i}>{i}</Text>)}
      </ScrollView>
      <Animated.View style={translateStyle}>
        <TextInput style={styles.input} />
      </Animated.View>
    </View>
  );
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    display: 'flex',
    justifyContent: 'space-between',
  },
  scroll: {
    width: '100%',
  },
  input: {
    height: 100,
    width: '100%',
    backgroundColor: 'green',
    borderWidth: 1,
  },
});

```

</details>

### CODE REVIEW

File: apple/keyboardObserver/REAKeyboardEventObserver.mm [hunk 1]
```diff
@@ -19,6 +20,9 @@ @implementation REAKeyboardEventObserver {
   NSMutableDictionary *_listeners;
   READisplayLink *_displayLink;
   KeyboardState _state;
+  CFTimeInterval _animtionStartTimestamp;
```
Comment (tjzel) [inline apple/keyboardObserver/REAKeyboardEventObserver.mm:23]:
```suggestion
  CFTimeInterval _animationStartTimestamp;
```

File: apple/keyboardObserver/REAKeyboardEventObserver.mm [hunk 2]
```diff
@@ -78,31 +82,108 @@ - (void)unsubscribeFromKeyboardEvents:(int)listenerId
...
+  timeProgress = fmax(fmin(timeProgress, 1), 0);
+  float x = timeProgress;
+  float progress = 1 - a1 * pow(1 - x, a2) - b1 * x * pow(1 - x, b2) - c1 * pow(x, 2) * pow(1 - x, c2);
+  return progress;
+}
+
+- (CGFloat)estimateCurrentKeyboardHeightDuringAppearing
+{
+  // Values comes from estimation: https://www.desmos.com/calculator/clhzejf5bs
+  float progress = [self estimateProgressForDuration:0.5 a1:1 a2:5.1 b1:1.6 b2:7.6 c1:0.2 c2:2.4];
```
Comment (tjzel) [inline apple/keyboardObserver/REAKeyboardEventObserver.mm:123]: Let's extract those values to some constants so we don't have to use them mutliple times like this.

File: apple/keyboardObserver/REAKeyboardEventObserver.mm [hunk 3]
```diff
@@ -78,31 +82,108 @@ - (void)unsubscribeFromKeyboardEvents:(int)listenerId
...
+  float currentKeyboardHeight = _targetKeyboardHeight * progress;
+  return currentKeyboardHeight;
+}
+
+- (CGFloat)estimateCurrentKeyboardHeightDuringDisappearing
+{
+  // Values comes from estimation: https://www.desmos.com/calculator/d3v550ofzs
+  float progress = [self estimateProgressForDuration:0.45 a1:1 a2:5.5 b1:2.5 b2:6.4 c1:1.6 c2:3.3];
```
Comment (tjzel) [inline apple/keyboardObserver/REAKeyboardEventObserver.mm:131]: Let's extract those values to some constants so we don't have to use them mutliple times like this.

File: apple/keyboardObserver/REAKeyboardEventObserver.mm [hunk 4]
```diff
@@ -117,17 +198,23 @@ - (void)keyboardWillChangeFrame:(NSNotification *)notification
...
   }
-
-  _measuringView.frame = CGRectMake(0, -1, 0, beginHeight);
-  [UIView animateWithDuration:animationDuration
-                   animations:^{
-                     self->_measuringView.frame = CGRectMake(0, -1, 0, endHeight);
-                   }];
-  [self runUpdater];
+  auto keyboardView = [self getKeyboardView];
+  bool hasKeyboardAniamtion = [self hasAnyAnimation:keyboardView];
```
Comment (tjzel) [inline apple/keyboardObserver/REAKeyboardEventObserver.mm:207]:
```suggestion
  bool hasKeyboardAnimation = [self hasAnyAnimation:keyboardView];
```

File: apple/keyboardObserver/REAKeyboardEventObserver.mm [hunk 5]
```diff
@@ -175,6 +262,36 @@ - (void)cleanupListeners
...
+
+// Inspired by:
+// https://stackoverflow.com/questions/32598490/show-uiview-with-buttons-over-keyboard-like-in-skype-viber-messengers-swift-i
+- (REAUIView *_Nullable)getKeyboardView
+{
+  if (_keyboardView) {
+    return _keyboardView;
+  }
+  NSArray<UIWindow *> *windows = [UIApplication sharedApplication].windows;
+  auto widow = [self findClass:@"UITextEffectsWindow" inViewsList:windows];
```
Comment (tjzel) [inline apple/keyboardObserver/REAKeyboardEventObserver.mm:289]: Let remove some sadness from the code.
```suggestion
  auto window = [self findClass:@"UITextEffectsWindow" inViewsList:windows];
```

File: apple/keyboardObserver/REAKeyboardEventObserver.mm [hunk 6]
```diff
@@ -78,31 +82,108 @@ - (void)unsubscribeFromKeyboardEvents:(int)listenerId
...
-      isAnimatingKeyboardChange ? _measuringView.layer.presentationLayer.frame : _measuringView.frame;
+  CFTimeInterval elapsedTime = _displayLink.targetTimestamp - _animtionStartTimestamp;
+  float timeProgress = elapsedTime / keyboardAnimationDuration;
+  timeProgress = fmax(fmin(timeProgress, 1), 0);
+  float x = timeProgress;
+  float progress = 1 - a1 * pow(1 - x, a2) - b1 * x * pow(1 - x, b2) - c1 * pow(x, 2) * pow(1 - x, c2);
+  return progress;
+}
+
+- (CGFloat)estimateCurrentKeyboardHeightDuringAppearing
```
Comment (tjzel) [inline apple/keyboardObserver/REAKeyboardEventObserver.mm:120]: How about?
```suggestion
- (CGFloat)estimateAppearingKeyboardHeight
```

File: apple/keyboardObserver/REAKeyboardEventObserver.mm [hunk 7]
```diff
@@ -78,31 +82,108 @@ - (void)unsubscribeFromKeyboardEvents:(int)listenerId
...
+{
+  // Values comes from estimation: https://www.desmos.com/calculator/clhzejf5bs
+  float progress = [self estimateProgressForDuration:0.5 a1:1 a2:5.1 b1:1.6 b2:7.6 c1:0.2 c2:2.4];
+  float currentKeyboardHeight = _targetKeyboardHeight * progress;
+  return currentKeyboardHeight;
+}
+
+- (CGFloat)estimateCurrentKeyboardHeightDuringDisappearing
```
Comment (tjzel) [inline apple/keyboardObserver/REAKeyboardEventObserver.mm:128]: How about?
```suggestion
- (CGFloat)estimateDisappearingKeyboardHeight
```

**Review by tjzel — APPROVED** [review]
All looks good to me, but I don't really know much about this part of iOS Reanimated. Please make sure that those changes work when you rapidly switch focus between TextInputs.

**Comment by piaskowyk** [comment]
I can confirm that, the focus switching between two inputs works correctly.

**Review by tomekzaw — APPROVED** [review]

**Comment by RohovDmytro** [comment]
A legend was born.

And then did this pull request.

Thanks.

**Pull request merged**